### ISSUE TITLE
token-cli: --blockhash does nothing

### ISSUE THREAD

**Issue by 2501babe** [solana-labs/solana-program-library#3442]
this is a regression introduced when the cli was converted to async. `BlockhashQuery` is a client wrapper that can be instantiated with a static blockhash which it always returns, a client that fetches a blockhash, or a static blockhash that it verifies before returning. it was removed for the nonblocking client and an unconditional call to `get_latest_blockhash`

i found this by accident when dumping a transaction, and im opening an issue instead of fixing it because:
* do we even need this? im not sure if there is a usecase for producing offline non-nonce transactions considering theyre good for like 30 seconds
* a quick fix here is pointless because if we decide to keep supporting it im going to have to work it into the client pr im working on anyway

@joncinque thoughts?

**Comment by joncinque** [comment]
Thanks for writing this up, I definitely broke this :sweat: 

We definitely need this, since even offline nonce transactions require a blockhash provided from the outside.  Check out the example and writeup here: https://spl.solana.com/token#example-offline-signing-with-multisig

Let's fix this the right way, by adding a nonblocking version of `BlockhashQuery` https://github.com/solana-labs/solana/blob/master/client/src/blockhash_query.rs -- since the implementation is pretty small, and a lot of the functions are deprecated, I think we're fine mostly copy-pasting with `async` and `await` wherever needed.

Alternatively, if you want to do it the DRY way, you can reuse the runtime from within the blocking RPC client in the blocking BlockhashQuery, and on the inside call the nonblocking `BlockhashQuery`, ie https://github.com/solana-labs/solana/blob/b6d38aad6979e7aca8fd431904af3cb90d87ebec/client/src/tpu_client.rs#L141-L146

But that seems like overkill.  What do you think?

**Comment by 2501babe** [comment]
cooool, im gonna look into copying the nonblocking client approach for a nonblocking `BlockhashQuery` then

this is another thing we are gonna have to pass in to `create_mint` et al so maybe we should have the complicated interface that the cli needs and then a wrapper with defaults (no nonce/offline stuff, at least) for normal consumers and expose them both...

**Comment by joncinque** [comment]
Hm, might be worth implementing that no-op client to make this a bit simpler... I can get a draft PR up with it to show what I mean.

**Comment by 2501babe** [comment]
```
#[derive(Debug, PartialEq, Eq)]
pub enum BlockhashQuery {
    None(Hash),
    FeeCalculator(Source, Hash),
    All(Source),
}
```
should we rename these? all the fee calculator stuff is going away so the only difference in functionality now is:
* returns a static hash unconditionally
* validates the static hash before returning it
* rpc passthrough
none/fee/all referred to which of the hash/fee pair it fetched. maybe `Static` `Validated` `Rpc`?

**Comment by 2501babe** [comment]
update, i landed this: https://github.com/solana-labs/solana/pull/27040 in the monorepo to provide a nonblocking blockhash client. its been backported to 1.11 so once we have a next release of that i can update versions and do this

**Comment by Poletgole** [comment]
> this is a regression introduced when the cli was converted to async. `BlockhashQuery` is a client wrapper that can be instantiated with a static blockhash which it always returns, a client that fetches a blockhash, or a static blockhash that it verifies before returning. it was removed for the nonblocking client and an unconditional call to `get_latest_blockhash`
> 
> i found this by accident when dumping a transaction, and im opening an issue instead of fixing it because:
> 
> * do we even need this? im not sure if there is a usecase for producing offline non-nonce transactions considering theyre good for like 30 seconds
> * a quick fix here is pointless because if we decide to keep supporting it im going to have to work it into the client pr im working on anyway
> 
> @joncinque thoughts?

If i may suggest a solution it will be talking to a customer agent through live chat https://direct.lc.chat/14390013/ they got the best, straightforward and  transparent solution to most programming blockchain related issues

**Comment by 2501babe** [comment]
fixed by #3449
